File: regions/system/library/dart/config/lib/src/vault.dart
import 'types.dart';

/// Vault から取得したシークレットで設定値を上書きする。
Config mergeVaultSecrets(Config config, Map<String, String> secrets) {
  if (secrets.containsKey('database.password') && config.database != null) {
    config.database!.password = secrets['database.password']!;
  }
  if (secrets.containsKey('redis.password') && config.redis != null) {
    config.redis!.password = secrets['redis.password'];
  }
  if (secrets.containsKey('kafka.sasl.username') &&
      config.kafka != null &&
      config.kafka!.sasl != null) {
    config.kafka!.sasl!.username = secrets['kafka.sasl.username']!;
  }
  if (secrets.containsKey('kafka.sasl.password') &&
      config.kafka != null &&
      config.kafka!.sasl != null) {
    config.kafka!.sasl!.password = secrets['kafka.sasl.password']!;
  }
  if (secrets.containsKey('redis_session.password') &&
      config.redisSession != null) {
    config.redisSession!.password = secrets['redis_session.password'];
  }
  if (secrets.containsKey('auth.oidc.client_secret') &&
      config.auth.oidc != null) {
    config.auth.oidc!.clientSecret = secrets['auth.oidc.client_secret'];
  }
  return config;
}

File: regions/system/library/rust/config/src/vault.rs
use std::collections::HashMap;

use crate::Config;

/// Vault シークレットで設定値を上書きする。
/// 異なるハッシャーを持つ `HashMap` にも対応するため、型パラメータを汎化する。
pub fn merge_vault_secrets<S: std::hash::BuildHasher>(config: &mut Config, secrets: &HashMap<String, String, S>) {
    if let Some(v) = secrets.get("database.password") {
        if let Some(ref mut db) = config.database {
            // clone_from はアロケーション再利用のため clone よりも効率的
            db.password.clone_from(v);
        }
    }
    if let Some(v) = secrets.get("redis.password") {
        if let Some(ref mut redis) = config.redis {
            redis.password = Some(v.clone());
        }
    }
    if let Some(v) = secrets.get("kafka.sasl.username") {
        if let Some(ref mut kafka) = config.kafka {
            if let Some(ref mut sasl) = kafka.sasl {
                // clone_from はアロケーション再利用のため clone よりも効率的
                sasl.username.clone_from(v);
            }
        }
    }
    if let Some(v) = secrets.get("kafka.sasl.password") {
        if let Some(ref mut kafka) = config.kafka {
            if let Some(ref mut sasl) = kafka.sasl {
                // clone_from はアロケーション再利用のため clone よりも効率的
                sasl.password.clone_from(v);
            }
        }
    }
    if let Some(v) = secrets.get("redis_session.password") {
        if let Some(ref mut redis_session) = config.redis_session {
            redis_session.password = Some(v.clone());
        }
    }
    if let Some(v) = secrets.get("auth.oidc.client_secret") {
        if let Some(ref mut oidc) = config.auth.oidc {
            oidc.client_secret = Some(v.clone());
        }
    }
}
